src/monitor/node_active_protocol.c

```diff
@@ -1108,6 +1108,22 @@ RemoveNode(AutoFailoverNode *currentNode)
 			formation->number_sync_standbys,
 			formation->formationId,
 			countSyncStandbys);
+
+		AutoFailoverNode *primaryNode =
+			GetPrimaryNodeInGroup(currentNode->formationId,
+								  currentNode->groupId);
+
+		if (primaryNode)
+		{
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to apply_settings "
+				"after removing standby node %s:%d from formation %s.",
+				primaryNode->nodeHost, primaryNode->nodePort,
+				currentNode->nodeHost, currentNode->nodePort,
+				formation->formationId);
+			SetNodeGoalState(primaryNode, REPLICATION_STATE_APPLY_SETTINGS, message);
```

Comment:
Can this have the opposite problem, where the primary node's goal state is overwritten by `ProceedGroupState` without anything actually happening? @soumyadeep2007 and I are wondering if perhaps `ProceedGroupState` needs to _know_ that a node has been dropped, somehow.

Reply:
I was wondering this too -- I did replace the `ProceedGroupState` call with this one, and all the tests passed, but I went with Dimitri's advice for the push up. I'm not sure about the test coverage.

Reply:
So, `synchronous_standby_names` is computed on the monitor in a function of the same name at https://github.com/citusdata/pg_auto_failover/blob/master/src/monitor/node_active_protocol.c#L2127. This function is called from the client side implementation of the “node-active protocol” from the function `fsm_apply_settings` at https://github.com/citusdata/pg_auto_failover/blob/master/src/bin/pg_autoctl/fsm_transition.c#L626. Without surprise it's called in the transition from PRIMARY to APPLY_SETTINGS already (see https://github.com/citusdata/pg_auto_failover/blob/master/src/bin/pg_autoctl/fsm.c#L320):

``` c
	{ PRIMARY_STATE, APPLY_SETTINGS_STATE, COMMENT_PRIMARY_TO_APPLY_SETTINGS, &fsm_apply_settings },
```

The `fsm_apply_settings` is also called from https://github.com/citusdata/pg_auto_failover/blob/master/src/bin/pg_autoctl/fsm_transition.c#L555 in the function `fsm_enable_sync_rep` which is used in more places in the FSM (https://github.com/citusdata/pg_auto_failover/blob/master/src/bin/pg_autoctl/fsm.c#L268):

``` c
	{ WAIT_PRIMARY_STATE, PRIMARY_STATE, COMMENT_WAIT_PRIMARY_TO_PRIMARY, &fsm_enable_sync_rep },
	{ JOIN_PRIMARY_STATE, PRIMARY_STATE, COMMENT_JOIN_PRIMARY_TO_PRIMARY, &fsm_enable_sync_rep },
```

Now, when removing a node from the group that was not the primary node, the primary node might be in any of SINGLE, PRIMARY, WAIT_PRIMARY, JOIN_PRIMARY, or APPLY_SETTINGS state. So a standby node might be removed during its WAIT_STANDBY to CATCHINGUP transition, say, allowing the primary to go from JOIN_PRIMARY to either WAIT_PRIMARY or PRIMARY. In that case, some transition will happen on the client side. 

I think we should only assign APPLY_SETTINGS when the primary node was already in the PRIMARY before the node removal, and still is in the PRIMARY state after the node removal. From there, we should probably extend that to “the state on the primary is the same after calling `ProceedGroupState`. If `ProceedGroupState` has assigned a state, that's the new state we should reach, that's it.

Now, to make the `synchronous_standby_names` change that is needed, we should extend the existing FSM transitions in the client side to call `fsm_apply_settings` where it's currently missing.